subxt/src/blocks/block_types.rs

```diff
@@ -0,0 +1,284 @@
+            Some(block) => block,
+            None => return Err(BlockError::block_hash_not_found(block_hash).into()),
+        };
+
+        Ok(BlockBody::new(
+            self.client.clone(),
+            block_details,
+            self.cached_events.clone(),
+        ))
+    }
+}
+
+/// The body of a block.
+pub struct BlockBody<T: Config, C> {
+    details: ChainBlockResponse<T>,
+    client: C,
+    cached_events: CachedEvents<T>,
+}
+
+impl<T, C> BlockBody<T, C>
+where
+    T: Config,
+    C: OfflineClientT<T>,
+{
+    pub(crate) fn new(
+        client: C,
+        details: ChainBlockResponse<T>,
+        cached_events: CachedEvents<T>,
+    ) -> Self {
+        Self {
+            details,
+            client,
+            cached_events,
+        }
+    }
+
+    /// Returns an iterator over the extrinsics in the block body.
+    pub fn extrinsics(&self) -> impl Iterator<Item = Extrinsic<'_, T, C>> {
+        self.details
+            .block
+            .extrinsics
+            .iter()
+            .enumerate()
+            .map(|(idx, e)| {
+                Extrinsic {
+                    index: idx as u32,
+                    bytes: &e.0,
+                    client: self.client.clone(),
+                    block_hash: self.details.block.header.hash(),
+                    cached_events: self.cached_events.clone(),
+                    _marker: std::marker::PhantomData,
+                }
+            })
+    }
+}
+
+/// A single extrinsic in a block.
+pub struct Extrinsic<'a, T: Config, C> {
+    index: u32,
+    bytes: &'a [u8],
+    client: C,
+    block_hash: T::Hash,
+    cached_events: Arc<AsyncMutex<Option<events::Events<T>>>>,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<'a, T, C> Extrinsic<'a, T, C>
+where
+    T: Config,
+    C: OfflineClientT<T>,
+{
+    /// The index of the extrinsic in the block.
+    pub fn index(&self) -> u32 {
+        self.index
+    }
+
+    /// The bytes of the extrinsic.
+    pub fn bytes(&self) -> &'a [u8] {
+        self.bytes
+    }
+}
+
+impl<'a, T, C> Extrinsic<'a, T, C>
+where
+    T: Config,
+    C: OnlineClientT<T>,
+{
+    /// The events associated with the extrinsic.
+    pub async fn events(&self) -> Result<ExtrinsicEvents<T>, Error> {
+        let events =
+            get_events(&self.client, self.block_hash, &self.cached_events).await?;
+        let ext_hash = T::Hashing::hash_of(&self.bytes);
+        Ok(ExtrinsicEvents::new(ext_hash, self.index, events))
+    }
+}
+
+/// The events associated with a given extrinsic.
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct ExtrinsicEvents<T: Config> {
+    // The hash of the extrinsic (handy to expose here because
+    // this type is returned from TxProgress things in the most
+    // basic flows, so it's the only place people can access it
+    // without complicating things for themselves).
+    ext_hash: T::Hash,
+    // The index of the extrinsic:
+    idx: u32,
```

Comment:
Could you add the idx getting function, because we want a combination of blockhash and ext index

Reply:
I don't see any harm in adding that (though note that you can get it from the extrinsic already)